src/transform-sdk/js/main.cc

```diff
@@ -315,8 +315,401 @@ std::expected<qjs::value, qjs::exception> make_write_event(
+                      "malformed schema def: bad reference: "
+                      "'subject' should be a string"));
+                }
+                auto vers = ref.get_property("version");
+                if (!vers.is_number()) {
+                    return std::unexpected(qjs::exception::make(
+                      ctx,
+                      "malformed schema def: bad reference: "
+                      "'version' should be a number"));
+                }
+                native_refs.emplace_back(
+                  std::string{name.string_data().view()},
+                  std::string{subj.string_data().view()},
+                  redpanda::sr::schema_version{
+                    static_cast<redpanda::sr::schema_version::type>(
+                      vers.as_integer())});
+            }
+        }
+
+        return redpanda::sr::schema{
+          std::string{raw_schema.string_data().view()},
+          static_cast<redpanda::sr::schema_format>(format.as_integer()),
+          std::move(native_refs)};
+    }
+
+    static std::expected<qjs::value, qjs::exception> make_subject_schema(
+      JSContext* ctx, const redpanda::sr::subject_schema& subj_schema) {
+        qjs::value obj = qjs::value::object(ctx);
+        std::expected<std::monostate, qjs::exception> result;
+        auto schema = make_schema(ctx, subj_schema.schema);
+        if (!schema.has_value()) {
+            return std::unexpected(schema.error());
+        }
+        result = obj.set_property("schema", schema.value());
+        if (!result.has_value()) {
+            return std::unexpected(result.error());
+        }
+        auto subject = qjs::value::string(ctx, subj_schema.subject);
+        if (!subject.has_value()) {
+            return std::unexpected(subject.error());
+        }
+        result = obj.set_property("subject", subject.value());
+        if (!result.has_value()) {
+            return std::unexpected(result.error());
+        }
+        result = obj.set_property(
+          "version", qjs::value::integer(ctx, subj_schema.version));
+        if (!result.has_value()) {
+            return std::unexpected(result.error());
+        }
+        result = obj.set_property(
+          "id", qjs::value::integer(ctx, subj_schema.id));
+        if (!result.has_value()) {
+            return std::unexpected(result.error());
+        }
+        return obj;
+    }
+
+    static std::expected<qjs::value, qjs::exception>
+    make_schema(JSContext* ctx, const redpanda::sr::schema& the_schema) {
+        qjs::value obj = qjs::value::object(ctx);
+        std::expected<std::monostate, qjs::exception> result;
+        auto schema_v = qjs::value::string(ctx, the_schema.raw_schema);
+        if (!schema_v.has_value()) {
+            return std::unexpected(schema_v.error());
+        }
+        result = obj.set_property("schema", schema_v.value());
+        if (!result.has_value()) [[unlikely]] {
+            return std::unexpected(result.error());
+        }
```

Comment:
Should we introduce the macros in the c++ SDK here? This is awful noisy.

Reply:
Yeah i keep flip flopping. Probably so

Reply:
leaned into `std::expected` a bit...little clunky in a couple of places, but overall not bad to use.

Reply:
seems like this is not controversial, closing the loop